ksql-engine/src/main/java/io/confluent/ksql/security/KsqlAccessValidatorCache.java

```diff
@@ -0,0 +1,147 @@
+
+    CacheKey(KsqlSecurityContext securityContext, String topicName, AclOperation operation) {
+      this.securityContext = securityContext;
+      this.topicName = topicName;
+      this.operation = operation;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == null || !(o instanceof CacheKey)) {
+        return false;
+      }
+
+      CacheKey other = (CacheKey)o;
+      return getuserName(securityContext).equals(getuserName(other.securityContext))
+          && topicName.equals(other.topicName)
+          && operation.code() == other.operation.code();
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(
+          getuserName(securityContext),
+          topicName,
+          operation.code()
+      );
+    }
+
+    private String getuserName(KsqlSecurityContext securityContext) {
+      return (securityContext.getUserPrincipal().isPresent())
+          ? securityContext.getUserPrincipal().get().getName()
+          : UNKNOWN_USER;
+    }
+  }
+
+  static class CacheValue {
+    private final boolean response;
+    private final RuntimeException exception;
+
+    CacheValue(boolean response, RuntimeException exception) {
+      this.response = response;
+      this.exception = exception;
+    }
+  }
+
+  private final LoadingCache<CacheKey, CacheValue> cache;
+  private final KsqlAccessValidator backendValidator;
+
+  public KsqlAccessValidatorCache(KsqlConfig ksqlConfig, KsqlAccessValidator backendValidator) {
+    this(ksqlConfig, backendValidator, Ticker.systemTicker());
+  }
+
+  @VisibleForTesting
+  KsqlAccessValidatorCache(
+      KsqlConfig ksqlConfig,
+      KsqlAccessValidator backendValidator,
+      Ticker cacheTicker
+  ) {
+    this.backendValidator = backendValidator;
+
+    long expiryTime = ksqlConfig.getLong(KsqlConfig.KSQL_AUTH_CACHE_EXPIRY_TIME);
+    long maxEntries = ksqlConfig.getLong(KsqlConfig.KSQL_AUTH_CACHE_MAX_ENTRIES);
+
+    cache = CacheBuilder.newBuilder()
+        .expireAfterWrite(expiryTime, TimeUnit.SECONDS)
+        .maximumSize(maxEntries)
+        .ticker(cacheTicker)
+        .build(buildCacheLoader());
+  }
+
+  private CacheLoader<CacheKey, CacheValue> buildCacheLoader() {
+    return new CacheLoader<CacheKey, CacheValue>() {
+      @Override
+      public CacheValue load(CacheKey cacheKey) {
+        try {
+          backendValidator.checkAccess(
+              cacheKey.securityContext,
+              cacheKey.topicName,
+              cacheKey.operation
+          );
+        } catch (KsqlTopicAuthorizationException e) {
+          return new CacheValue(DENIED, e);
+        }
+
+        return new CacheValue(ALLOWED, null);
+      }
+    };
+  }
+
+  @Override
+  public void checkAccess(
+      KsqlSecurityContext securityContext,
+      String topicName,
+      AclOperation operation
+  ) {
+    CacheKey cacheKey = new CacheKey(securityContext, topicName, operation);
+    CacheValue cacheValue = cache.getUnchecked(cacheKey);
```

Comment:
again for my understanding, any unchecked exception will propagate to the caller of `checkAccess` ?

Reply:
Yes. there is a unit test that catches this named shouldThrowExceptionWhenBackendValidatorThrowsAnException on KsqlCacheAccessValidatorTest